src/auth/auth-config.ts

```diff
@@ -0,0 +1,576 @@
+      rpEntityId: this.rpEntityId,
+      callbackURL: this.callbackURL,
+      enableRequestSigning: this.enableRequestSigning,
+    };
+  }
+}
+
+/**
+ * Defines the OIDCConfig class used to convert a client side configuration to its
+ * server side representation.
+ */
+export class OIDCConfig implements OIDCAuthProviderConfig {
+  public readonly enabled: boolean;
+  public readonly displayName?: string;
+  public readonly providerId: string;
+  public readonly issuer: string;
+  public readonly clientId: string;
+
+  /**
+   * Static method convert a client side request to a OIDCConfigServerRequest.
+   * Throws an error if validation fails. If the request is not a OIDCConfig request,
+   * returns null.
+   *
+   * @param {OIDCAuthProviderRequest} options The options object to convert to a server request.
+   * @param {boolean=} ignoreMissingFields Whether to ignore missing fields.
+   * @return {?OIDCConfigServerRequest} The resulting server request or null if not valid.
+   */
+  public static buildServerRequest(
+      options: OIDCAuthProviderRequest,
+      ignoreMissingFields: boolean = false): OIDCConfigServerRequest | null {
+    const makeRequest = typeof options !== 'undefined' &&
+        validator.isNonNullObject(options) &&
+        (options.providerId || ignoreMissingFields);
+    if (!makeRequest) {
+      return null;
+    }
+    const request: OIDCConfigServerRequest = {};
+    // Validate options.
+    OIDCConfig.validate(options, ignoreMissingFields);
+    request.enabled = options.enabled;
+    request.displayName = options.displayName;
+    request.issuer = options.issuer;
+    request.clientId = options.clientId;
+    return request;
+  }
+
+  /**
+   * Returns the provider ID corresponding to the resource name if available.
+   *
+   * @param {string} resourceName The server side resource name
+   * @return {?string} The provider ID corresponding to the resource, null otherwise.
+   */
+  public static getProviderIdFromResourceName(resourceName: string): string | null {
+    // name is of form projects/project1/oauthIdpConfigs/providerId1
+    const matchProviderRes = resourceName.match(/\/oauthIdpConfigs\/(oidc\..*)$/);
+    if (!matchProviderRes || matchProviderRes.length < 2) {
+      return null;
+    }
+    return matchProviderRes[1];
+  }
+
+  /**
+   * @param {any} providerId The provider ID to check.
+   * @return {boolean} Whether the provider ID corresponds to an OIDC provider.
+   */
+  public static isProviderId(providerId: any): boolean {
+    return validator.isNonEmptyString(providerId) && providerId.indexOf('oidc.') === 0;
+  }
+
+  /**
+   * Validates the OIDCConfig options object. Throws an error on failure.
+   *
+   * @param {OIDCAuthProviderRequest} options The options object to validate.
+   * @param {boolean=} ignoreMissingFields Whether to ignore missing fields.
+   */
+  public static validate(options: OIDCAuthProviderRequest, ignoreMissingFields: boolean = false) {
```

Comment:
This PR is a lot of code, and it's not easy to figure out which should be public and which not. Lets make sure we are only exposing things that we really need.